### PR TITLE
Handle indexers in nested object initializers

### PR DESCRIPTION
**Pull request by 333fred**
When I removed the usage of `Lazy<T>` from IOperation, I mishandled setting whether a property initializer in a nested object initializer expression is a object initializer itself. This meant that during the build of the parent node, we'd correctly check the setter for the property to ensure that it was valid, but then when handling deriving the arguments for the child node we'd use the wrong method (the getter, instead of the setter). Practically, this only had observably bad results from indexers: indexers without a get method would cause the code to null ref. This is because when deriving the arguments for the getter for non-indexed properties, we'd use the getter, and then immediately bail out because the getter had no arguments. Fixes https://github.com/dotnet/roslyn/issues/45692.

Commit-by-commit review is recommended. Commit 1 nullable enables code so that I could figure out what the bug was. The second commit actually fixes the bug.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred It looks like there are legitimate CI failures.

**Comment by 333fred** [comment]
@AlekseyTs should be addressed now.

@dotnet/roslyn-compiler for reviews please.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs [hunk 1]
```diff
@@ -331,6 +335,8 @@ internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containing
                 return ImmutableArray<IArgumentOperation>.Empty;
             }
 
+            Debug.Assert(optionalParametersMethod is not null);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:338]: >Debug.Assert(optionalParametersMethod is not null); [](start = 12, length = 51)

Would it be better (more robust) to simply check `optionalParametersMethod is null || ...` in the `if` above?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:338]: It would certainly be less likely to crash, but it's likely that would be masking a bug, which could then result in downstream bugs in analyzers.

---
In reply to: [497823219](https://github.com/dotnet/roslyn/pull/48168#discussion_r497823219) [](ancestors = 497823219)

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs [hunk 2]
```diff
@@ -208,18 +208,20 @@ internal IOperation CreateMemberInitializerInitializedMember(BoundNode initializ
...
             switch (containingExpression.Kind)
             {
                 case BoundKind.ObjectInitializerMember:
                     {
                         var boundObjectInitializerMember = (BoundObjectInitializerMember)containingExpression;
-                        var property = (PropertySymbol)boundObjectInitializerMember.MemberSymbol;
-                        MethodSymbol accessor = isObjectOrCollectionInitializer ? property.GetOwnOrInheritedGetMethod() : property.GetOwnOrInheritedSetMethod();
+                        var property = (PropertySymbol?)boundObjectInitializerMember.MemberSymbol;
+                        MethodSymbol? accessor = isObjectOrCollectionInitializer ? property.GetOwnOrInheritedGetMethod() : property.GetOwnOrInheritedSetMethod();
+                        Debug.Assert(property is not null);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:221]: >Debug.Assert(property is not null); [](start = 24, length = 35)

Should the assert be moved right after the declaration of ```property```? It looks like it is dereferenced in the previous statement.

File: src/Compilers/CSharp/Portable/Symbols/PropertySymbolExtensions.cs [hunk 3]
```diff
@@ -14,12 +12,12 @@ internal static class PropertySymbolExtensions
         /// If the property has a GetMethod, return that.  Otherwise check the overridden
         /// property, if any.  Repeat for each overridden property.
         /// </summary>
-        public static MethodSymbol GetOwnOrInheritedGetMethod(this PropertySymbol property)
+        public static MethodSymbol? GetOwnOrInheritedGetMethod(this PropertySymbol? property)
         {
-            while ((object)property != null)
+            while (property is not null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/PropertySymbolExtensions.cs:17]: >is not [](start = 28, length = 6)

It looks like this PR doesn't follow the general guideline to keep unrelated refactorings out of PRs that change product behavior. Also, in general we prefer avoiding making stylistic changes to existing code. Everybody has personal preferences and if analyzer doesn't flag the code, it is in a good state style/formatting wise.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/PropertySymbolExtensions.cs:17]: Since this is an ask mode fix, I would prefer changes like that reverted and keeping this PR to a necessary minimum to address the issue.

---
In reply to: [497930116](https://github.com/dotnet/roslyn/pull/48168#discussion_r497930116) [](ancestors = 497930116)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

File: src/Compilers/CSharp/Portable/Symbols/PropertySymbolExtensions.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/PropertySymbolExtensions.cs:17]: These changes are what allowed me to figure out the bug here. My strategy was to nullable enable the callstack, which allowed me to figure out where the bug was coming from. I'd prefer to keep them.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:221]: No. Those are extension methods that accept a `null` input. I could certainly move it, but it doesn't particularly matter.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:221]: > Those are extension methods that accept a null input. I could certainly move it, but it doesn't particularly matter.

I think that would benefit the reader.

---
In reply to: [497941233](https://github.com/dotnet/roslyn/pull/48168#discussion_r497941233) [](ancestors = 497941233)

File: src/Compilers/CSharp/Portable/Symbols/PropertySymbolExtensions.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/PropertySymbolExtensions.cs:17]: > These changes are what allowed me to figure out the bug here. My strategy was to nullable enable the callstack, which allowed me to figure out where the bug was coming from. I'd prefer to keep them.

Changes to comparisons to null and the like are unnecessary. I think they should be reverted and any other changes that are not meant to change product behavior. Nullable annotations and asserts are fine, of course.

---
In reply to: [497940563](https://github.com/dotnet/roslyn/pull/48168#discussion_r497940563) [](ancestors = 497940563)

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 4)
